docs/_build/man/fcust.1
.\" Man page generated from reStructuredText.
.
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.TH "FCUST" "1" "Nov 21, 2021" "0.1.1" "Folder Custodian"
.SH NAME
fcust \- Folder Custodian Documentation
.SH FOLDER CUSTODIAN
\fI\%\fP\fI\%\fP\fI\%Documentation Status\fP\fI\%Updates\fP
.sp
Linux Common Folder Custodian
.INDENT 0.0
.IP \(bu 2
Free software: GNU General Public License v3
.IP \(bu 2
Documentation:  \fI\%Read the Docs\fP
.UNINDENT
.SS Why?
.sp
The family computer runs Fedora, has many users and we want to share some files with each other.
To achive this we create a shared common folder.
This package solves the problems that come with our use case.
.SS Features
.sp
This package is intended to perform maintenance over a folder shared across many users in a Fedora Workstation.
The folder should belong to a group for which all users are members.
.INDENT 0.0
.IP \(bu 2
Go through a folder’s content and enforce common group ownership.
.IP \(bu 2
Go through a folder’s content and enforce common group read and write permissions as
well as set groupid sticky bit.
.IP \(bu 2
Provide a command line command with sane defaults for enforcing common folder group ownership
and permissions.
.IP \(bu 2
Write logs to \fB/tmp/fcust/$USER.log\fP\&.
.IP \(bu 2
Provides user systemd service to periodically enforce common folder group ownership
and permissions. Runs on user log out.
.UNINDENT
.sp
The key problem this package intends to solve is that files moved from one location to another do not inherit
proper permissions in the commonly used folder and therefore when another user tries to access them they encounter
problems.
.SS Quick Start Guide
.INDENT 0.0
.IP \(bu 2
Install fcust:
.INDENT 2.0
.INDENT 3.5
.sp
.nf
.ft C
$ sudo dnf copr enable iolaum/fcust
$ sudo dnf install fcust
.ft P
.fi
.UNINDENT
.UNINDENT
.IP \(bu 2
Manually create a common group and add as members the users that will have access to it.
For example:
.INDENT 2.0
.INDENT 3.5
.sp
.nf
.ft C
$ sudo groupadd family
$ sudo usermod \-a \-G family user1
$ sudo usermod \-a \-G family user1
.ft P
.fi
.UNINDENT
.UNINDENT
.IP \(bu 2
Create that folder and assign proper permissions:
.INDENT 2.0
.INDENT 3.5
.sp
.nf
.ft C
$ mkdir /path/to/common/folder/
$ chown :family /path/to/common/folder/
$ chmod 2775 /path/to/common/folder/
.ft P
.fi
.UNINDENT
.UNINDENT
.IP \(bu 2
Setup fcust:
.INDENT 2.0
.INDENT 3.5
.sp
.nf
.ft C
$ fcust setup /path/to/common/folder/
$ fcust activate
.ft P
.fi
.UNINDENT
.UNINDENT
.IP \(bu 2
In order to run fcust manually on a properly permissioned common folder run:
.INDENT 2.0
.INDENT 3.5
.sp
.nf
.ft C
$ fcust run /path/to/common/folder/
.ft P
.fi
.UNINDENT
.UNINDENT
.UNINDENT
.SH INSTALLATION
.SS Fedora COPR Repositories
.sp
The recommended way to install Folder Custodian is through the Fedora Copr repositories:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
$ sudo dnf copr enable iolaum/fcust
$ sudo dnf install fcust
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
This is the preferred method to install Folder Custodian,
as it will always install the most recent stable release.
.SS Python Packaging Index
.sp
Alternatively Folder Custodian can be installed as a python package:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
$ pip install fcust
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
If you don’t have \fI\%pip\fP installed, this \fI\%Python installation guide\fP can guide
you through the process.
.SS From sources
.sp
The sources for Folder Custodian can be downloaded from the \fI\%Github repo\fP\&.
.sp
For more information follow the Contributing page.
.SH FCUST
.SS fcust package
.SS Submodules
.SS fcust.cli module
.sp
Console script for fcust.
.SS fcust.fcust module
.sp
Main module.
.INDENT 0.0
.TP
.B class fcust.fcust.CommonFolder(folder_path: pathlib.PosixPath, common_group: str = \(aq\(aq)
Bases: \fBobject\fP
.sp
Main class regarding management of a folder that is commonly used across many users.
.INDENT 7.0
.IP \(bu 2
The class creates a logfile at /tmp/fcust/$USER.log
.IP \(bu 2
By default the class assumes that the root of the common folder is configured
correctly and then tries to enforce appropriate permissions.
.UNINDENT
.INDENT 7.0
.TP
.B Parameters
.INDENT 7.0
.IP \(bu 2
\fBfolder_path\fP – Path where the common folder is located.
.IP \(bu 2
\fBcommon_group\fP – Group name regarding the common folder.
If not passed the existing group of the folder will be assumed to be the proper folder.
.UNINDENT
.UNINDENT
.INDENT 7.0
.TP
.B enforce_permissions()
We read the contents of a specified directory and enforce unix permissions.
.sp
Files should have 664 permissions
Folders should have 2775 permisions (ie also setguid bit)
Group should be common golder’s group.
.sp
The function only changes permissions if the user is owner of the relevant resource.
This is done to avoid the need for root access, but requires the function to be
run by all the users sharing the common folder.
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B fcust.fcust.create_logger(cgroup: str, lpath: str = \(aq/tmp/fcust/\(aq)
Function that creates a modified logger to cater fcust functionality.
.INDENT 7.0
.TP
.B Parameters
\fBcgroup\fP – group ownership of the main directory
.UNINDENT
.UNINDENT
.SS fcust.service module
.sp
Utilities for Folder Custodian Service.
.INDENT 0.0
.TP
.B fcust.service.activate_service()
After a fcust setup is run we need to activate the service we installed.
.UNINDENT
.INDENT 0.0
.TP
.B fcust.service.create_fcust_service_unit(folder_path: pathlib.PosixPath, unit_path: pathlib.PosixPath)
Create a systemd user unit for folder cutodian.
Use predefined template and modify where needed.
.sp
We want the service to run when the user logs out so that all the changes
they made are fixed if needed. We consult the following sources
to create the appropriate systemd service template:
.INDENT 7.0
.IP \(bu 2
\fI\%https://wiki.archlinux.org/index.php/Systemd/User\fP
.IP \(bu 2
\fI\%https://superuser.com/questions/1037466/\fP
how\-to\-start\-a\-systemd\-service\-after\-user\-login\-and\-stop\-it\-before\-user\-logout/1269158
.IP \(bu 2
\fI\%https://askubuntu.com/questions/293312/\fP
execute\-a\-script\-upon\-logout\-reboot\-shutdown\-in\-ubuntu/796157#796157
.UNINDENT
.INDENT 7.0
.TP
.B Parameters
.INDENT 7.0
.IP \(bu 2
\fBfolder_path\fP – Path where the common folder is located.
.IP \(bu 2
\fBunit_path\fP – Path where the common folder is located.
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B fcust.service.create_user_unit_path(create_folder: bool = False)
We want to add a systemd user unit to run fcust on certain times. In order to do
this we want to crete a unit at a proper location. According to:
\fI\%https://wiki.archlinux.org/index.php/Systemd/User\fP
our options are:
.INDENT 7.0
.IP \(bu 2
/usr/lib/systemd/user/:
.sp
where units provided by installed packages belong.
.IP \(bu 2
~/.local/share/systemd/user/
.sp
where units of packages that have been installed in the home directory belong.
.IP \(bu 2
/etc/systemd/user/
.sp
where system\-wide user units are placed by the system administrator.
.IP \(bu 2
~/.config/systemd/user/
.sp
where the user puts their own units.
.UNINDENT
.sp
We opt to use the latter choice.
.sp
This function constructs the proper systemd user unit path where it will be installed.
It also creates the necessary folder if it doesn’t exist.
.INDENT 7.0
.TP
.B Parameters
\fBcreate_folder\fP – If true create the folder that the unit service will be installed.
.TP
.B Returns
Path object for the location the unit service will be installed.
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B fcust.service.deactivate_service()
Deactivate a running fcust service.
.UNINDENT
.SS Module contents
.sp
Top\-level package for Folder Custodian.
.SH CONTRIBUTING
.sp
Contributions are welcome, and they are greatly appreciated! Every little bit
helps, and credit will always be given.
.sp
You can contribute in many ways:
.SS Types of Contributions
.SS Report Bugs
.sp
Report bugs at the project’s \fI\%issues page\fP\&.
.sp
If you are reporting a bug, please include:
.INDENT 0.0
.IP \(bu 2
Your operating system name and version.
.IP \(bu 2
Any details about your local setup that might be helpful in troubleshooting.
.IP \(bu 2
Detailed steps to reproduce the bug.
.UNINDENT
.SS Fix Bugs
.sp
Look through the GitHub issues for bugs. Anything tagged with “bug” and “help
wanted” is open to whoever wants to implement it.
.SS Implement Features
.sp
Look through the GitHub issues for features. Anything tagged with “enhancement”
and “help wanted” is open to whoever wants to implement it.
.SS Write Documentation
.sp
Folder Custodian could always use more documentation, whether as part of the
official Folder Custodian docs, in docstrings, or even on the web in blog posts,
articles, and such.
.SS Submit Feedback
.sp
The best way to send feedback is to file an issue at the project’s \fI\%issues page\fP\&.
.sp
If you are proposing a feature:
.INDENT 0.0
.IP \(bu 2
Explain in detail how it would work.
.IP \(bu 2
Keep the scope as narrow as possible, to make it easier to implement.
.IP \(bu 2
Remember that this is a volunteer\-driven project, and that contributions
are welcome :)
.UNINDENT
.SS Get Started!
.sp
Ready to contribute? Here’s how to set up \fIfcust\fP for local development.
.INDENT 0.0
.IP 1. 3
Fork the \fIfcust\fP repo on GitHub.
.IP 2. 3
Install Required packages.
.INDENT 3.0
.INDENT 3.5
.sp
.nf
.ft C
$ sudo dnf install python3\-pip python3\-tox python3\-wheel make git findutils hadolint
.ft P
.fi
.UNINDENT
.UNINDENT
.IP 3. 3
Clone your fork locally.
.INDENT 3.0
.INDENT 3.5
.sp
.nf
.ft C
$ git clone git@github.com:your_name_here/fcust.git
$ cd fcust
.ft P
.fi
.UNINDENT
.UNINDENT
.IP 4. 3
From the root of the repository create a python virtual environment to use for our project.
Install the project in editable mode along with developer dependencies.
.INDENT 3.0
.INDENT 3.5
.sp
.nf
.ft C
$ python3 \-m venv venv
$ source venv/bin/activate
(venv) $ pip install \-e .[dev]
.ft P
.fi
.UNINDENT
.UNINDENT
.IP 5. 3
Create a branch for local development.
.INDENT 3.0
.INDENT 3.5
.sp
.nf
.ft C
$ git checkout \-b name\-of\-your\-bugfix\-or\-feature
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Now you can make your changes locally.
.IP 6. 3
When you’re done making changes, check that your changes pass code quality checks
and tests.
.INDENT 3.0
.INDENT 3.5
.sp
.nf
.ft C
$ make code
$ make test
.ft P
.fi
.UNINDENT
.UNINDENT
.IP 7. 3
Commit your changes and push your branch to GitHub.
.INDENT 3.0
.INDENT 3.5
.sp
.nf
.ft C
$ git add .
$ git commit \-m "Your detailed description of your changes."
$ git push origin name\-of\-your\-bugfix\-or\-feature
.ft P
.fi
.UNINDENT
.UNINDENT
.IP 8. 3
Submit a pull request through the GitHub website.
.UNINDENT
.SS Pull Request Guidelines
.sp
Before you submit a pull request, check that it meets these guidelines:
.INDENT 0.0
.IP 1. 3
The pull request should include tests.
.IP 2. 3
If the pull request adds functionality, the docs should be updated. Put
your new functionality into a function with a docstring, and add the
feature to the list in README.rst.
.IP 3. 3
The pull request should work for Python 3.9, for PyPy and Fedora. Check
the \fI\%relevant GitHub Actions page\fP and make sure that the tests pass.
.UNINDENT
.SS Developing on Fedora Silverblue
.sp
If developing on Fedora Silverblue the following set up is suggested:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
# Crete new toolbox
$ toolbox create dev
# Add toolbox entry to /etc/hosts to avoid
# warning: Could not canonicalize hostname: toolbox
$ sudo nano /etc/hosts
# 127.0.0.1 ... toolbox
$ toolbox enter dev
# Following commands are inside the toolbox
$ sudo groupadd family
$ sudo usermod \-a \-G family $USER
# exit and re\-enter toolbox container for changes to take effect.
$ cd $source_code_root_repository
$ sudo dnf install make fedpkg python3\-wheel python3\-devel python3\-sphinx python3\-click hadolint
$ make code
$ make test
$ make fedpkg
.ft P
.fi
.UNINDENT
.UNINDENT
.SS Deploying
.sp
A reminder for the maintainers on how to deploy.
Make sure all your changes are committed.
Then run:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
$ bump2version patch # possible: major / minor / patch
$ git push
$ git push \-\-tags
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Then make sure to update \fIHISTORY.rst\fP and the \fIfcust.spec\fP changelog section. Afterwards
commit and push to Github.
.sp
If all checks pass manually deploy to PyPI. Merge to the release branch to automatically trigger a COPR build.
.SH CREDITS
.SS Development Lead
.INDENT 0.0
.IP \(bu 2
Nikolaos Perrakis <\fI\%nikperrakis@gmail.com\fP>
.UNINDENT
.SS Contributors
.sp
This package was initalised with \fI\%Cookiecutter\fP and the \fI\%audreyr/cookiecutter\-pypackage\fP project template.
.SH HISTORY
.SS 0.1.1 (2020\-11\-22)
.INDENT 0.0
.IP \(bu 2
Added ability to see recent service logs.
.IP \(bu 2
Releasing beta version of feature complete package.
.UNINDENT
.SS 0.0.14 (2020\-11\-10)
.INDENT 0.0
.IP \(bu 2
Adding functionality to create, start and stop systemd user service
.UNINDENT
.SS 0.0.10 (2020\-11\-07)
.INDENT 0.0
.IP \(bu 2
F33 release
.IP \(bu 2
Minor Documentation update
.UNINDENT
.SS 0.0.9 (2020\-10\-13)
.INDENT 0.0
.IP \(bu 2
Hotfixed log access path bug
.UNINDENT
.SS 0.0.8 (2020\-10\-10)
.INDENT 0.0
.IP \(bu 2
Draft packaging as a fedora 32 rpm package
.UNINDENT
.SS 0.0.5 (2020\-10\-07)
.INDENT 0.0
.IP \(bu 2
Added command line command for folder contents permission enformcement
.IP \(bu 2
Added documentation
.UNINDENT
.SS 0.0.3 (2020\-10\-05)
.INDENT 0.0
.IP \(bu 2
Alpha version with common folder enformcement functionality.
.UNINDENT
.INDENT 0.0
.IP \(bu 2
genindex
.IP \(bu 2
modindex
.IP \(bu 2
search
.UNINDENT
.SH AUTHOR
Nikolaos Perrakis
.SH COPYRIGHT
2020, Nikolaos Perrakis
.\" Generated by docutils manpage writer.
.


HISTORY.rst
=======
History
=======

1.0.0 (2021-11-21)
-------------------

* Updated package for Fedora 35.
* Releasing production version.

0.1.1 (2020-11-22)
-------------------

* Added ability to see recent service logs.
* Releasing beta version of feature complete package.

0.0.14 (2020-11-10)
-------------------

* Adding functionality to create, start and stop systemd user service

0.0.10 (2020-11-07)
-------------------

* F33 release
* Minor Documentation update

0.0.9 (2020-10-13)
------------------

* Hotfixed log access path bug

0.0.8 (2020-10-10)
------------------

* Draft packaging as a fedora 32 rpm package

0.0.5 (2020-10-07)
------------------

* Added command line command for folder contents permission enformcement
* Added documentation

0.0.3 (2020-10-05)
------------------

* Alpha version with common folder enformcement functionality.


fcust/fcust.py
"""
Main module.
"""

from pathlib import PosixPath, Path
from shutil import chown
from pwd import getpwuid
from os import getuid
import logging


class CommonFolder:
    """

    Main class regarding management of a folder that is commonly used across many users.

    :param folder_path: Path where the common folder is located.
    :param common_group: Group name regarding the common folder.
      If not passed the existing group of the folder will be assumed to be the proper folder.
    """

    def __init__(self, folder_path: PosixPath, common_group: str = ""):
        """
        Initialize CommonFolder class.
        """

        if not isinstance(folder_path, PosixPath):
            raise TypeError(f"Expected PosixPath object instead of {type(folder_path)}")

        if not folder_path.exists():
            raise FileNotFoundError(
                "Folder is expected to be present when the class is initialized."
            )

        self.path: PosixPath = folder_path

        if common_group == "":
            # TODO: Add warning when logging gets added.
            self.group: str = self.path.group()
        else:
            self.group = common_group

        self.user: str = getpwuid(getuid()).pw_name

        # create logger
        logger = logging.getLogger("fcust")
        logger.setLevel(logging.DEBUG)
        # Create handlers
        sh = logging.StreamHandler()
        # TODO: make filename depend on day/time?
        fh = logging.FileHandler("/tmp/fcust.log", mode="w")
        sh.setLevel(logging.DEBUG)
        fh.setLevel(logging.DEBUG)
        # Create formatters and add it to handlers
        format1 = logging.Formatter("%(name)s - %(levelname)s - %(message)s")
        format2 = logging.Formatter(
            "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
        )
        sh.setFormatter(format1)
        fh.setFormatter(format2)
        # Add handlers to the logger
        logger.addHandler(sh)
        logger.addHandler(fh)
        # add logger object to class
        self.logger: logging.Logger = logger

        logger.info(f"CommonFolder class instantiated for: {str(folder_path)}")

    def enforce_permissions(self):
        """
        We read the contents of a specified directory and enforce unix permissions.

        Files should have 664 permissions
        Folders should have 2775 permisions (ie also setguid bit)
        Group should be common golder's group.

        The function only changes permissions if the user is owner of the relevant resource.
        This is done to avoid the need for root access, but requires the function to be
        run by all the users sharing the common folder.
        """

        try:
            self.logger.info("Starting permissions enforcement")
            # Let's iterate over folders we find first - glob returns a generator!
            list1 = self.path.glob("**/*")
            for il in list1:
                pil = Path(il)

                # Fix group membership if needed
                if pil.group() != self.group and pil.owner() == self.user:
                    chown(pil, group=self.group)
                    self.logger.debug(f"Fixing group for {str(pil)}")

                # Check folder permissions and fix if needed
                if pil.is_dir():
                    perms: str = oct(pil.stat().st_mode)[-4:]
                    if perms != "2775" and pil.owner() == self.user:
                        pil.chmod(0o42775)
                        self.logger.debug(f"Fixing permissions for {str(pil)}")

                # Check folder permissions and fix if needed
                if pil.is_file():
                    perms: str = oct(pil.stat().st_mode)[-4:]
                    if perms != "0664" and pil.owner() == self.user:
                        pil.chmod(0o40664)
                        self.logger.debug(f"Fixing permissions for {str(pil)}")

        except Exception as exc:
            self.logger.critical(exc, exc_info=True)

        self.logger.info("Finished permissions enforcement")
